File: individual-name-matching/python-examples/append-simkeys-to-file.py
#!/usr/bin/env python3
"""
Simple batch example calling Interzoid's Get Full Name Match API.

This script:
  1. Reads one name per line from an input text file.
  2. URL-encodes each name.
  3. Calls the API for each input.
  4. Writes a CSV file with two columns: original value and SimKey.

Only the Python standard library is used.
"""

import csv
import json
import urllib.request
import urllib.parse
import urllib.error

# Replace this with your API key from: https://www.interzoid.com/manage-api-account
API_KEY = "YOUR_API_KEY_HERE"

# Hardcoded input and output file names
INPUT_FILE_NAME = "sample-input-file.txt"
OUTPUT_FILE_NAME = "output.csv"


def main():
    """
    Main program logic.
    """

    # Try opening input file for reading.
    try:
        in_file = open(INPUT_FILE_NAME, "r", encoding="utf-8")
    except OSError as e:
        print("Error opening input file:", e)
        return

    with in_file:
        # Try opening output CSV for writing.
        try:
            out_file = open(OUTPUT_FILE_NAME, "w", newline="", encoding="utf-8")
        except OSError as e:
            print("Error creating output file:", e)
            return

        with out_file:
            csv_writer = csv.writer(out_file)

            line_number = 0

            # Iterate over the input file line by line.
            for line in in_file:
                line_number += 1

                # Remove only trailing newline, not other spaces.
                original_value = line.rstrip("\n")

                # Skip empty lines (optional).
                if len(original_value) == 0:
                    continue

                # URL encode the name.
                name_param = urllib.parse.quote(original_value)

                # Build API URL.
                url = (
                    "https://api.interzoid.com/getfullnamematch"
                    f"?license={API_KEY}"
                    f"&fullname={name_param}"
                )

                # Call the API and read response.
                try:
                    with urllib.request.urlopen(url) as response:
                        if response.status != 200:
                            print(
                                f"API error on line {line_number} "
                                f"({original_value!r}): HTTP {response.status}"
                            )
                            csv_writer.writerow([original_value, ""])
                            continue

                        # Read raw body string.
                        body = response.read().decode("utf-8")

                except urllib.error.URLError as e:
                    print(
                        f"API error on line {line_number} "
                        f"({original_value!r}): {e}"
                    )
                    csv_writer.writerow([original_value, ""])
                    continue

                # Parse the JSON response.
                try:
                    data = json.loads(body)
                except json.JSONDecodeError as e:
                    print(
                        f"JSON parse error for {original_value!r}: {e}"
                    )
                    csv_writer.writerow([original_value, ""])
                    continue

                # Extract SimKey field from response.
                sim_key = data.get("SimKey", "")

                # Write a CSV row with original value + SimKey.
                try:
                    csv_writer.writerow([original_value, sim_key])
                except OSError as e:
                    print(
                        f"CSV write error for {original_value!r}: {e}"
                    )
                    continue

    print("Done. Results written to", OUTPUT_FILE_NAME)


if __name__ == "__main__":
    main()

File: individual-name-matching/go-examples/append-simkeys-to-file.go
package main

import (
	"bufio"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
)

// Replace this with your API key from https://www.interzoid.com/manage-api-account
const API_KEY = "YOUR_API_KEY_HERE"

// Hardcoded input and output file names
const inputFileName = "sample-input-file.txt"
const outputFileName = "output.csv"

func main() {
	// Open the input file
	inFile, err := os.Open(inputFileName)
	if err != nil {
		fmt.Println("Error opening input file:", err)
		return
	}
	defer inFile.Close()

	// Create the output CSV file
	outFile, err := os.Create(outputFileName)
	if err != nil {
		fmt.Println("Error creating output file:", err)
		return
	}
	defer outFile.Close()

	csvWriter := csv.NewWriter(outFile)
	defer csvWriter.Flush()

	// JSON mapping struct
	type Response struct {
		SimKey  string `json:"SimKey"`
		Code    string `json:"Code"`
		Credits string `json:"Credits"`
	}

	scanner := bufio.NewScanner(inFile)
	lineNumber := 0

	for scanner.Scan() {
		lineNumber++
		originalValue := scanner.Text()

		// Skip blank lines optionally
		if len(originalValue) == 0 {
			continue
		}

		// URL encode the name
		nameParam := url.QueryEscape(originalValue)

		// Build API URL
		apiURL := "https://api.interzoid.com/getfullnamematch?license=" + API_KEY +
			"&fullname=" + nameParam

		// Make request
		resp, err := http.Get(apiURL)
		if err != nil {
			fmt.Printf("API error on line %d (%q): %v\n", lineNumber, originalValue, err)
			_ = csvWriter.Write([]string{originalValue, ""})
			continue
		}

		body, err := ioutil.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			fmt.Printf("Response read error for %q: %v\n", originalValue, err)
			_ = csvWriter.Write([]string{originalValue, ""})
			continue
		}

		var result Response
		if err := json.Unmarshal(body, &result); err != nil {
			fmt.Printf("JSON parse error for %q: %v\n", originalValue, err)
			_ = csvWriter.Write([]string{originalValue, ""})
			continue
		}

		// Write one row per input line
		err = csvWriter.Write([]string{originalValue, result.SimKey})
		if err != nil {
			fmt.Printf("CSV write error for %q: %v\n", originalValue, err)
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading input file:", err)
		return
	}

	fmt.Println("Done. Results written to", outputFileName)
}
